### ISSUE TITLE
Optimize likelihood cache

### ISSUE THREAD

**Issue by hyanwong** [tskit-dev/tsdate#247]
At the moment we cache the likelihoods, which are a function of edge span and number of mutations above the edge (along with distance between timepoints). However, there may be many combinations of `(span, #mutations)` which are unique, and hence don't need caching.

We could probably work out which combinations are unique or not easily enough:

```python
# assume we have pre-calculated the mutations above each edge and stored them in `mut_edges`
mut_span, counts = np.unique(
    np.vstack((mut_edges, ts.edges_right-ts_edges_left)),
    axis=1,
    return_counts=True,
)
for num_muts, span in mut_span[:, counts > 1].T:
    # this combination of num_muts and span is duplicated, so cache the likelihoods for the different time intervals
```

**Comment by nspope** [comment]
Is it still worth precomputing these for unique cases, because then it's possible to parallelize?

**Comment by hyanwong** [comment]
I don't think this is relevant any more, is it @nspope? We don't cache likelihoods in the variational_gamma method. So I'm closing it.
